src/EditorFeatures/Core/RenameTracking/RenameTrackingTaggerProvider.RenameTrackingCommitter.cs

```diff
@@ -251,25 +269,27 @@ private void UpdateWorkspaceForResetOfTypedIdentifier(Workspace workspace, Solut
                 localUndoTransaction.AddUndo(undoPrimitiveBefore);
 
                 if (!workspace.TryApplyChanges(newSolution))
-                {
-                    Contract.Fail("Rename Tracking could not update solution.");
-                }
+                    return (NotificationSeverity.Error, EditorFeaturesResources.Rename_operation_could_not_complete_due_to_external_change_to_workspace);
 
-                if (!_refactorNotifyServices.TryOnAfterGlobalSymbolRenamed(workspace, changedDocuments, symbol, newName, throwOnFailure: false))
+                try
                 {
-                    var notificationService = workspace.Services.GetService<INotificationService>();
-                    notificationService.SendNotification(
-                        EditorFeaturesResources.Rename_operation_was_not_properly_completed_Some_file_might_not_have_been_updated,
-                        EditorFeaturesResources.Rename_Symbol,
-                        NotificationSeverity.Information);
+                    if (!_refactorNotifyServices.TryOnAfterGlobalSymbolRenamed(workspace, changedDocuments, symbol, newName, throwOnFailure: false))
+                        return (NotificationSeverity.Information, EditorFeaturesResources.Rename_operation_was_not_properly_completed_Some_file_might_not_have_been_updated);
+
+                    return null;
                 }
+                finally
+                {
+                    // If we successfully update the workspace then make sure the undo transaction is committed and is
+                    // always able to undo anything any other external listener did.
```

Comment:
same concept as before.  if the workspace application fails, we don't add these.  but if it succeeds, we always add this, regardless of waht our external-notification system does.